gymnasium/envs/registration.py

```diff
@@ -32,63 +30,118 @@
 else:
     import importlib.metadata as metadata
 
-if sys.version_info >= (3, 8):
-    from typing import Literal
+if sys.version_info < (3, 8):
+    from typing_extensions import Protocol
 else:
-    from typing_extensions import Literal
-
-from gymnasium import Env, error, logger
+    from typing import Protocol
 
 
 ENV_ID_RE = re.compile(
     r"^(?:(?P<namespace>[\w:-]+)\/)?(?:(?P<name>[\w:.-]+?))(?:-v(?P<version>\d+))?$"
 )
 
 
-def load(name: str) -> Callable:
-    """Loads an environment with name and returns an environment creation function.
+__all__ = [
+    "EnvSpec",
+    "registry",
+    "current_namespace",
+    # Core Functions
+    "register",
+    "make",
+    "spec",
+    "pprint_registry",
+    # Minor Functions
+    "namespace",
+    "get_env_id",
+    "parse_env_id",
+    "find_highest_version",
+    "load_env",
+    "load_plugin_envs",
+]
 
-    Args:
-        name: The environment name
 
-    Returns:
-        Calls the environment constructor
+class EnvCreator(Protocol):
+    """Function type expected for an environment."""
+
+    def __call__(self, **kwargs: Any) -> Env:
+        ...
+
+
+@dataclass
+class EnvSpec:
+    """A specification for creating environments with :meth:`gymnasium.make`.
+
+    * **id**: The string used to create the environment with :meth:`gymnasium.make`
+    * **entry_point**: A string for the environment location, ``(import path):(environment name)`` or a function that creates the environment.
+    * **reward_threshold**: The reward threshold for completing the environment.
+    * **nondeterministic**: If the observation of an environment cannot be repeated with the same initial state, random number generator state and actions.
+    * **max_episode_steps**: The max number of steps that the environment can take before truncation
+    * **order_enforce**: If to enforce the order of :meth:`gymnasium.Env.reset` before :meth:`gymnasium.Env.step` and :meth:`gymnasium.Env.render` functions
+    * **autoreset**: If to automatically reset the environment on episode end
+    * **disable_env_checker**: If to disable the environment checker wrapper in :meth:`gymnasium.make`, by default False (runs the environment checker)
+    * **kwargs**: Additional keyword arguments passed to the environment during initialisation
```

Comment:
If we do reorder arguments, the docstring should be adapted to that

Reply:
Agreed